### PR TITLE
Fix a bug in hyperexpand

### PR DESCRIPTION
**Pull request by certik**
- [x] " I have asked Sergey privately whether his is ok with this arrangement, going forward. And if he is, I also intend to ask people in the SymPy community whether they are ok with this arrangement."

### CODE REVIEW

**Comment by asmeurer** [comment]
Sergey gave permission on the mailing list to use this patch without appending the LICENSE.

**Comment by gxyd** [comment]
I read the discussion on mailing list, just out of curiosity, i have seen in past that after Sergey created his separate `omg` CAS, he was still using the commits made by some of authors against sympy master, and incorporated them into his own branch (though he had till made changes to his copyright). The thing is now he incorporated those commits from authors sympy repo to his repo without asking, but to incorporate his changes here permission is required! (why?). Was it ethically (or legally or whatever) wrong, how Sergey incorporated those changes?
I am asking just out of curiosity. I have no intention of getting into this matter.

**Comment by certik** [comment]
@gxyd good point. My understanding is that if you take a BSD licensed code, say, from SymPy, you just need to copy the license. But you are allowed to add more people or entities into the copyright, I think that's fine and I've seen it recommended online (e.g. http://programmers.stackexchange.com/a/158011/152863), though Sergey should make it clear that the patches coming from SymPy only have the copyright:
```
Copyright (c) 2006-2016 SymPy Development Team
```
While patches contributing by him have the copyright:
```
Copyright (c) 2006-2015 SymPy Development Team,
              2015-2016 Sergey B Kirpichev 
```
One should not be removing people or entities from the copyright, which is the situation if we want to get code back from him. It is asymmetric, as you noted.

**Comment by certik** [comment]
Btw, Sergey has also removed the AUTHORS file, thus removing the actual copyright holders, e.g. members of the "SymPy Development Team", though he is keeping the credits in `docs/aboutus.rst`. Strictly speaking, it is complying with the license, since it only requires to "retain the above copyright notice", which he did. It also makes sense to just maintain one list, though the list in `doc/aboutus.rst` is highly out of date, it's really hard to keep that current. The most accurate and up to date is git log.

It is indeed interesting to first remove all SymPy's author names from the top level directory (thus the whole top level directory has just Sergey's name in the LICENSE file), and then require that Sergey's name is included in the SymPy's LICENSE copyright file in the same top level directory in SymPy (and also require not to be in SymPy's AUTHORS).

But this is Sergey's decision, I think he has a right to do that, I think he is following the license, so that's what matters (and he does keep the authors at least in `doc/aboutus.rst`). What matters for us is how we attribute people and if we follow the license terms.

But, just to be sure, I have asked Sergey privately whether his is ok with this arrangement, going forward. And if he is, I also intend to ask people in the SymPy community whether they are ok with this arrangement.

**Comment by moorepants** [comment]
I think this is fine and it allows us to take other patches from his fork as needed.

**Comment by gxyd** [comment]
This got forgotten. I will merge in 24 hrs if no one objects.

**Comment by asmeurer** [comment]
@gxyd you should wait for Ondrej to give the OK on this one.

**Comment by shivamvats** [comment]
@certik Do we want to merge this?

**Comment by certik** [comment]
I am against merging this until we get an explicit permission from Sergey, that this arrangement is fine with him. I haven't got such a permission yet.

**Comment by certik** [comment]
Francesco Bonazzi wrote here: https://groups.google.com/d/msg/sympy/MBF5a1IY6pc/HRVj2pBOBgAJ

"Sergey wrote me that we should feel free to port anything from diofant provided that we keep the copyright notice and obey to the LICENSE terms."

So I think this is giving us explicit permission to use his work, and I think this PR preserves the copyright notice and obeys the license terms. 

So I am +1 to merge this.

File: LICENSE [hunk 1]
```diff
@@ -26,3 +26,38 @@ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
...
 OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 DAMAGE.
+
+--------------------------------------------------------------------------------
+
+Patches by Sergey B Kirpichev <skirpichev@gmail.com> merged after January 1,
+2016 are licensed as:
+
+Copyright (c) 2006-2015 SymPy Development Team,
+              2015-2016 Sergey B Kirpichev
```
Comment (isuruf) [inline LICENSE]: This has been changed to,
```
Copyright (c) 2006-2016 SymPy Development Team,
              2013-2017 Sergey B Kirpichev
```

Comment (isuruf) [inline LICENSE]: Given that there are patches from 2017 in diofant, this really should have been,
```
Copyright (c) 2006-2017 SymPy Development Team,
              2013-2017 Sergey B Kirpichev
```
to comply with SymPy's license

**Comment by certik** [comment]
IMPORTANT: I am withdrawing my +1. Please do not merge this.

Sergey just replied to my email from 2/5/2016, He did NOT give me an explicit permission to use his patches. He said he gave me permission to quote the first sentence of his reply, here it is: "I'm not planning to discuss anything related to my open-source projects privately."

In this situation I am against merging any of his patches without his explicit permission.

**Comment by certik** [comment]
I have asked for a permission publicly at https://github.com/diofant/diofant/issues/432:
```
Hi Sergey,

Due to popular demand, people are asking if SymPy can merge some of your patches back to SymPy. I have created a PR (https://github.com/sympy/sympy/pull/10531) which copies your BSD license into SymPy, and starts using your patches. The arrangement would be that in the future, we can just merge patches from your "master" branch, after that PR is merged.

I would like to get your explicit approval that we can do that. Are you ok with this arrangement?

Thanks,
Ondrej
```

**Comment by certik** [comment]
Sergey has answered in https://github.com/diofant/diofant/issues/432, and I think that makes it clear that we can use his code in SymPy, as long as we obey the BSD license, as we do in this PR.

There is one minor point regarding the licensing of Diofant, that I raised here:

https://github.com/diofant/diofant/issues/445

But I think otherwise the situation is clear. After discussing with few people, I am fine with merging this PR, and then go ahead and merge other improvements that Sergey has done.

**Comment by jksuom** [comment]
Maybe the polynomial test should be improved before merging. In its current form it seems to reject all hypergeometric polynomials with only one nonpositive integer upper parameter if there are also such lower parameters. For example,
```
In [13]: h = hyper((-1, 1), (-2,), x)

In [14]: hyperexpand(h)
Out[14]: ∞
```
However, this h is the linear polynomial `1 + x/2` from the (truncated) hypergeometric series.

**Comment by certik** [comment]
Ok, I updated the license based on Sergey's feedback and he approved the current wording (ebfac73f3be02138d13bd10fbdefa22e78b982ad) in https://github.com/diofant/diofant/issues/445#issuecomment-281722971, so we have reached an agreement and the administrative part is done as far as I am concerned.

@jksuom do you know how to fix that?

**Comment by jksuom** [comment]
Yes, I did send a PR  (#12182) when it looked like this was not going to be merged. The fix has been merged in https://github.com/skirpichev/diofant/commit/92d1d998a32d1114e1728396c7c8b5ca81a678ff

**Comment by certik** [comment]
@jksuom thanks. I've ported the fix you mentioned and the test. Anything else to be done?

**Comment by jksuom** [comment]
Hypergeometric functions with common nonpositive integer upper and lower parameters are not well defined. There are (at least) two different ways of handling them. This PR will produce a polynomial:
```
>>> h = hyper((-1, 1), (-1,), x)
>>> hyperexpand(h)
x + 1
```
while my PR, which begins by eliminating common parameters, creates a series (or a polynomial of higher degree in some cases):
```
 >>> h = hyper((-1, 1), (-1,), x)
>>> hyperexpand(h)
1/(-x + 1)
```
The reason for this difference is that the limit of `hyper((a, 1), (c,), x)` at `(a, c) == (-1, -1)` does not exist in the two dimensional plane. It only exists if `(a, c)` is restricted to a line passing through `(-1, -1)`, and then the value of the limit depend on the choice of the line.

Another point of view is to consider `hyper` as a solution of the hypergeometric differential equation. In general, `hyper` is defined as the (unique) solution with initial condition `h(0) = 1`. In the generic case, there is only one solution, but with parameters `a, b, c = -1, 1, -1` there are two linearly independent ones. So there is no preferred definition of `hyper((-1, 1), (-1,), x)`.

The behaviour of mpmath depends on the keyword `eliminate`. It is True by default, and then the result is as in the latter example above. That is why I chose that method. But I think the other choice is equally valid. It is probably possible to have both methods implemented in another PR.

**Comment by certik** [comment]
@jksuom thanks. We should probably implement both.

In order to move forward, do you want to merge this PR, or do you want to first implement the other method (by porting it from your other PR), or do something else? I am fine either way, but if you are also ok with either way, I vote to merge this PR, so that we can move on, and then improve later.

**Comment by jksuom** [comment]
I think this could be merged now and the keyword `eliminate` could be implemented later.

**Comment by certik** [comment]
@jksuom it's in! Thanks everybody for the review.

**Pull request merged**